lib/obj.js

```diff
@@ -1,365 +1,447 @@
-            var materials = result.materials;
-            var images = result.images;
-            return processObj(objFile, info, materials, images);
-        });
+function Node() {
+    this.name = undefined;
+    this.meshes = [];
 }
 
-function processObj(objFile, info, materials, images) {
-    return new Promise(function(resolve) {
-        // A vertex is specified by indexes into each of the attribute arrays,
-        // but these indexes may be different. This maps the separate indexes to a single index.
-        var vertexCache = {};
-        var vertexCount = 0;
-
-        var vertexArray = [];
-
-        var positions = [];
-        var normals = [];
-        var uvs = [];
-
-        var positionMin = [Number.MAX_VALUE, Number.MAX_VALUE, Number.MAX_VALUE];
-        var positionMax = [-Number.MAX_VALUE, -Number.MAX_VALUE, -Number.MAX_VALUE];
-
-        var hasNormals = info.hasNormals;
-        var hasUVs = info.hasUVs;
-
-        var materialGroups = {}; // Map material to index array
-        var currentIndexArray;
-
-        // Switch to the material-specific index array, or create it if it doesn't exist
-        function useMaterial(material) {
-            if (!defined(materials[material])) {
-                useDefaultMaterial();
-            } else {
-                currentIndexArray = materialGroups[material];
-                if (!defined(currentIndexArray)) {
-                    currentIndexArray = [];
-                    materialGroups[material] = currentIndexArray;
-                }
-            }
-        }
+function Mesh() {
+    this.name = undefined;
+    this.primitives = [];
+    this.positions = new ArrayStorage(ComponentDatatype.FLOAT);
+    this.normals = new ArrayStorage(ComponentDatatype.FLOAT);
+    this.uvs = new ArrayStorage(ComponentDatatype.FLOAT);
+}
+
+function Primitive() {
+    this.material = undefined;
+    this.indices = new ArrayStorage(ComponentDatatype.UNSIGNED_INT);
+}
 
-        function useDefaultMaterial() {
-            var defaultMaterial = 'czmDefaultMat';
-            if (!defined(materials[defaultMaterial])) {
-                materials[defaultMaterial] = Material.getDefault();
+// OBJ regex patterns are modified from ThreeJS (https://github.com/mrdoob/three.js/blob/master/examples/js/loaders/OBJLoader.js)
+var vertexPattern = /v( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)/;                                                      // v float float float
+var normalPattern = /vn( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)/;                                                     // vn float float float
+var uvPattern = /vt( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)/;                                                                               // vt float float
+var facePattern1 = /f( +-?\d+)\/?( +-?\d+)\/?( +-?\d+)\/?( +-?\d+)?\/?/;                                                                        // f vertex vertex vertex ...
```

Comment:
I'm pretty inexperienced with regexes, so I could have used a note that the `facePattern` regexes return something like `[generatedVertexID, position, texCoord, Normal]`